harness/determined/common/storage/boto3_credential_manager.py

```diff
@@ -0,0 +1,135 @@
+import logging
+import os
+import threading
+import time
+from typing import Any
+
+import boto3
+from botocore import session
+from botocore.compat import ensure_unicode
+from botocore.credentials import (
+    CredentialProvider,
+    Credentials,
+    ReadOnlyCredentials,
+    SharedCredentialProvider,
+)
+from botocore.session import get_session
+
+
+class RefreshableCredentialProvider(CredentialProvider):  # type: ignore
+    """
+    Creates a refreshable credential provider class given an existing credential provider in
+    the boto3 credential chain.
+
+    """
+
+    METHOD = "managed-refresh-cred"
+
+    def __init__(self, credential_provider: SharedCredentialProvider, check_every: int = 2) -> None:
+        super().__init__()
+        self.check_every = check_every
+        self.credential_provider = credential_provider
+
+    def load(self) -> Credentials:
+        return self.credential_provider.load() and RefreshableSharedCredentials(
+            credentials_provider=self.credential_provider, check_every=self.check_every
+        )
+
+
+class RefreshableSharedCredentials(Credentials):  # type: ignore
+    def __init__(
+        self,
+        check_every: int,
+        credentials_provider: SharedCredentialProvider,
+    ):
+        self._credentials_provider = credentials_provider
+        self._check_every = check_every
+        self._lock = threading.Lock()
+        self._check_time = time.time() + check_every
+        self._load_and_set_credentials()
+
+    def _load_and_set_credentials(self) -> None:
+        credentials = self._credentials_provider.load()
+        self._last_loaded = self._credentials_modified_time()
+        self.access_key = credentials.access_key
+        self.secret_key = credentials.secret_key
+        self.token = credentials.token
+        self._frozen_credentials = ReadOnlyCredentials(
+            credentials.access_key, credentials.secret_key, credentials.token
+        )
+
+    def _credentials_file(self) -> Any:
+        path = self._credentials_provider._creds_filename
+        path = os.path.expandvars(path)
+        path = os.path.expanduser(path)
+        return path
+
+    def _credentials_modified_time(self) -> float:
+        credentials_file = self._credentials_file()
+        return os.stat(credentials_file).st_mtime
+
+    def _refresh_needed(self) -> bool:
+        return self._credentials_modified_time() > self._last_loaded
```

Comment:
this is a really slight nit, but maybe return true if `self._credentials_modified_time` is different at all?  Probably would never matter I guess.